pkg/apiserver/rest/dualwriter_syncer_test.go

```diff
@@ -0,0 +1,238 @@
+				},
+				expectedOutcome: false,
+			},
+			{
+				name: "storage has an extra 1 entry (foo4)",
+				setupLegacyFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(legacyListWith3items, nil)
+				},
+				setupStorageFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(storageListWith4items, nil)
+					m.On("Delete", mock.Anything, "foo4", mock.Anything, mock.Anything).Return(exampleObj, false, nil)
+				},
+				expectedOutcome: true,
+			},
+			{
+				name: "storage has an extra 1 entry (foo4) - fail to delete",
+				setupLegacyFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(legacyListWith3items, nil)
+				},
+				setupStorageFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(storageListWith4items, nil)
+					m.On("Delete", mock.Anything, "foo4", mock.Anything, mock.Anything).Return(exampleObj, false, errors.New("error"))
+				},
+				expectedOutcome: false,
+			},
+			{
+				name: "storage is missing 1 entry (foo3) and has an extra 1 entry (foo4)",
+				setupLegacyFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(legacyListWith3items, nil)
+				},
+				setupStorageFn: func(m *mock.Mock) {
+					m.On("List", mock.Anything, mock.Anything).Return(storageListWith3itemsMissingFoo2, nil)
+					m.On("Update", mock.Anything, "foo2", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(exampleObj, false, nil)
+					m.On("Delete", mock.Anything, "foo4", mock.Anything, mock.Anything).Return(exampleObj, false, nil)
+				},
+				expectedOutcome: true,
+			},
+		}
+
+	// mode 1
+	for _, tt := range tests {
+		t.Run("Mode-1-"+tt.name, func(t *testing.T) {
+			l := (LegacyStorage)(nil)
+			s := (Storage)(nil)
+			lm := &mock.Mock{}
+			um := &mock.Mock{}
+
+			ls := legacyStoreMock{lm, l}
+			us := storageMock{um, s}
+
+			if tt.setupLegacyFn != nil {
+				tt.setupLegacyFn(lm)
+			}
+			if tt.setupStorageFn != nil {
+				tt.setupStorageFn(um)
+			}
+
+			outcome, err := legacyToUnifiedStorageDataSyncer(context.Background(), Mode1, ls, us, "test.kind", p, &fakeServerLock{}, &request.RequestInfo{})
+			if tt.wantErr {
+				assert.Error(t, err)
+				return
+			}
+
+			assert.NoError(t, err)
+			assert.Equal(t, tt.expectedOutcome, outcome)
+		})
+	}
+
+	// mode 2
+	for _, tt := range tests {
+		t.Run("Mode-2-"+tt.name, func(t *testing.T) {
+			l := (LegacyStorage)(nil)
+			s := (Storage)(nil)
+			lm := &mock.Mock{}
+			um := &mock.Mock{}
+
+			ls := legacyStoreMock{lm, l}
+			us := storageMock{um, s}
+
+			if tt.setupLegacyFn != nil {
+				tt.setupLegacyFn(lm)
+			}
+			if tt.setupStorageFn != nil {
+				tt.setupStorageFn(um)
+			}
+
+			outcome, err := legacyToUnifiedStorageDataSyncer(context.Background(), Mode1, ls, us, "test.kind", p, &fakeServerLock{}, &request.RequestInfo{})
```

Comment:
@filewalkwithme looks like this should be Mode2.

Reply:
You are right, I will fix this in a follow up PR right now